libraries/chain/controller.cpp

```diff
@@ -402,11 +402,13 @@ struct controller_impl {
       }
    }
 
-   void dmlog_applied_transaction(const transaction_trace_ptr& t) {
+   void dmlog_applied_transaction(const transaction_trace_ptr& t, const signed_transaction *trx = nullptr) {
       // dmlog_applied_transaction is called by push_scheduled_transaction
       // where transient transactions are not possible, and by push_transaction
       // only when the transaction is not transient
       if (auto dm_logger = get_deep_mind_logger(false)) {
+         if (trx && !trx->actions.empty() && trx->actions[0].name == "onblock"_n)
```

Comment:
Use `trace.cpp` `is_onblock`. No need to pass in the `signed_transaction`.
BTW, you would want to use `get_transaction()` of `packed_transaction` not `get_signed_transaction()`.

Reply:
Hum, should I also rewrite `dm_logger->on_onblock()` to take a transaction_trace?

Reply:
Yes, probably better than doing the `get<packed_transaction>(t.receipt.trx).get_signed_transaction()` with appropriate error checking. Also I missed that `on_onblock` was packing a `signed_transaction`, so ignore my comment about `get_transaction` vs `get_signed_transaction` as signed transaction is desired here apparently.

Reply:
Would that be better than passing the `signed_transaction` like today, but using `is_onblock` as you suggested?

Reply:
`t.receipt` is just a `transaction_receipt_header`, not a `transaction_receipt`, so it doesn't have the `trx` member.

Reply:
ugh, so it is.

Reply:
What about using `is_onblock()` like this? Or leave as is?

```
   void dmlog_applied_transaction(const transaction_trace_ptr& t, const signed_transaction *trx = nullptr) {
      // dmlog_applied_transaction is called by push_scheduled_transaction
      // where transient transactions are not possible, and by push_transaction
      // only when the transaction is not transient
      if (auto dm_logger = get_deep_mind_logger(false)) {
         if (is_onblock(*t))
            dm_logger->on_onblock(*trx);
         dm_logger->on_applied_transaction(self.head_block_num() + 1, t);
      }
   }
```

Reply:
Yes. `const signed_transaction* trx = nullptr`. Should probably check for or assert trx != nullptr.